### ISSUE TITLE
A list sorted in a custom way can result in the Combobox NOT working

### ISSUE THREAD

**Issue by FollowSteph** [vaadin/framework#9845]
This affects the latest version of Vaadin, specifically version 8.1.2. The WAR file is attached as a zipped file at the bottom of this bug report. If you run refresh the webpage from the WAR file it will randomly select the third item in the ComboBox although it should always select the third item.

In the example provided if the list of Cars is sorted according to the ID's then the ComboBox works as expected. However if I randomly shuffle the list of Cars then the combobox will randomly appear on the screen unselected (it's actually not really random but this is discussed later). 

```
        final VerticalLayout layout = new VerticalLayout();

        List<Car> cars = Arrays.asList(
                new Car(1, "Honda1"),
                new Car(2, "Honda2"),
                new Car(3, "Honda3"),
                new Car(4, "Honda4"),
                new Car(5, "Honda5"),
                new Car(6, "Honda6"),
                new Car(7, "Honda7"),
                new Car(8, "Honda8"),
                new Car(9, "Honda9"),
                new Car(10, "Honda10"),
                new Car(11, "Honda11")
        );

        // The shuffle will cause the ComboBox to be unselected randomly
        Collections.shuffle(cars);

        Person person = new Person("BuyerOfCar3", 3);

        ComboBox<Car> carComboBox = new ComboBox<Car>();
        carComboBox.setItemCaptionGenerator(Car::getName);
        carComboBox.setItems(cars);

        Binder<Person> binder = new Binder<Person>();

        binder.forField(carComboBox)
                .withConverter(new ComboBoxCarToIdConverter())
                .bind(Person::getCarID, Person::setCarID);

        binder.readBean(person);

        layout.addComponents(carComboBox);
        setContent(layout);
```

In the above code the line: Collections.shuffle(cars); will randomly cause the ComboBox to be unselected. If the list is sorted by ID then there is no issue however as soon as I randomly shuffle the list the ComboBox will randomly fail. In fact it's not randomly fail but rather it seems to search until the value is greater than (or something similar) and will stop checking in the list after that. In other words the Binder logic seems to assume the list is sorted in some fashion and will therefore fail for lists sorted in non-expected ways. In my real code I sort the list based on two criteria, one of which is the data's status (for example if the car is sold or not), and hence the ID's appear non-sequentially.

The ComboBoxCarToIdConverter is:

```
    class ComboBoxCarToIdConverter implements Converter<Car, Integer>
    {
        @Override
        public Result<Integer> convertToModel(Car car, ValueContext context)
        {
            if(car == null)
                return Result.ok(-1);
            return Result.ok(car.getId());
        }

        @Override
        public Car convertToPresentation(Integer id, ValueContext context)
        {
            if(id > -1)
                return new Car(id);
            return null;
        }
    }
```
 
And the Data objects are:

```
public class Person
{
	private String name;
	private int carID;

	public Person(String name, int carID)
	{
		this.name = name;
		this.carID = carID;
	}

	public String getName()
	{
		return name;
	}

	public void setName(String name)
	{
		this.name = name;
	}

	public int getCarID()
	{
		return carID;
	}

	public void setCarID(int carID)
	{
		this.carID = carID;
	}
}
```

And:

```
public class Car
{
	private int id;
	private String name;

	public Car(int id)
	{
		this.id = id;
	}

	public Car(int id, String name)
	{
		this.id = id;
		this.name = name;
	}

	public int getId()
	{
		return id;
	}

	public void setId(int id)
	{
		this.id = id;
	}

	public String getName()
	{
		return name;
	}

	public void setName(String name)
	{
		this.name = name;
	}

	@Override
	public int hashCode()
	{
		return getId();
	}

	@Override
	public boolean equals(Object obj)
	{
		if(this == obj)
			return true;
		if(obj == null || getClass() != obj.getClass())
			return false;

		return getId() == ((Car)obj).getId();
	}
}
```

[VaadinDebug-1.0-SNAPSHOT.zip](https://github.com/vaadin/framework/files/1236105/VaadinDebug-1.0-SNAPSHOT.zip)

**Comment by FollowSteph** [comment]
I'm just following up to see if anyone has had a chance to look into this bug. And if so if it's expected to be resolved. 

In case you're wondering about a use case it could for example if the list of cars is sorted first by status (with an icon) and then by name. So for example the status could be sold, backorder, instore, etc. We have several different use cases where the list is sorted according to a special order and all the comboboxes in these cases fail.

**Comment by tsuoanttila** [comment]
Hi @FollowSteph ,

The AbstractSingleSelect selection logic used in ComboBox currently seems to be a bit faulty when you have multiple different objects representing the same item. You should be able to fix the issue by overriding `DataProvider#getId` in the DataProvider you use to return the ID of the car. So instead of `ComboBox.setItems(cars)` you should create a new `ListDataProvider` and override the method `getId`. This will then map the object by the ID and not by the instance of the Car. Hope this helps with the issue for now. It is a known problem, but it's not trivial to solve for all possible corner cases for the ComboBox.

**Comment by FollowSteph** [comment]
I tried that bit it didn't change anything. The code I have is:

```
CarListDataProvider listDataProvider = new CarListDataProvider(tenants);
CarComboBox.setDataProvider(listDataProvider);

```
where

```
class CarListDataProvider extends CarDataProvider<Car>
{
	public CarListDataProvider(Collection<Car> items)
	{
		super(items);
	}

	@Override
	public Object getId(Car car)
	{
		return car.getId();
	}
}
```

And I had the exact same issue...

**Comment by FollowSteph** [comment]
Are any updates on this issue? Is this going to be fixed in the near future?

**Comment by TatuLund** [comment]
This issue is possibly fixed in Vaadin 8.5 by https://github.com/vaadin/framework/pull/10796
